### PR TITLE
Creating Freetrial option for newly registered users

### PR DESCRIPTION
**Pull request by sowmiyasubramaniyan**
This PR is about creating a free trial option for newly registered users
        1. Here I have created the Faveo cloud trial button.
        2. When a new user clicks that button it will create the invoice and order for the Trail product.
        3. And it will call the tenant API and create the domain for users.
        
enhancement : #1693

### CODE REVIEW

File: vendor/doctrine/deprecations/test_fixtures/vendor/doctrine/foo/Bar.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
```
Comment (sandesh556) [inline vendor/doctrine/deprecations/test_fixtures/vendor/doctrine/foo/Bar.php:1]: Remove this files from the PR

File: vendor/doctrine/deprecations/test_fixtures/vendor/doctrine/foo/Baz.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
```
Comment (sandesh556) [inline vendor/doctrine/deprecations/test_fixtures/vendor/doctrine/foo/Baz.php:1]: Remove this file from this PR

**Review by sandesh556 — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/FreeTrailController.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+use Illuminate\Http\Request;
+use Lang;
+
+class FreeTrailController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware('auth');
+        $invoice = new Invoice();
+        $this->invoice = $invoice;
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:26]: Avoiding unnecessary variable declaration. You can write $this->invoice = new Invoice();

File: app/Http/Controllers/FreeTrailController.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+        $invoiceItem = new InvoiceItem();
+        $this->invoiceItem = $invoiceItem;
+
+        $order = new Order();
+        $this->order = $order;
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:32]: As above mentioned

File: app/Http/Controllers/FreeTrailController.php [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+        $subscription = new Subscription();
+        $this->subscription = $subscription;
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:35]: remove variable

File: app/Http/Controllers/FreeTrailController.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+        $id = $request->get('id');
+        if (Auth::user()->id == $id) {
+            $user_login = User::find($id);
+            if ($user_login->first_time_login != 0) {
+                return errorResponse(Lang::get('message.false'), 400);
+            }
+            $user = User::where('id', $id)->update(['first_time_login' => 1]);
+
+            $invoice = $this->generateFreetrailInvoice();
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:52]: Please remove unused variable declaration $invoice,$user,$inoice_items

File: app/Http/Controllers/FreeTrailController.php [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+            $invoice_items = $this->createFreetrailInvoiceItems();
+
+            $order = $this->executeFreetrailOrder();
+
+            return successResponse(Lang::get('message.succs_fre'), $order, 200);
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:58]: refactor the code
return successResponse(trans('message.succs_fre'), $this->executeFreetrailOrder(), 200);

File: app/Http/Controllers/FreeTrailController.php [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        $tax_rule = new \App\Model\Payment\TaxOption();
+        $rule = $tax_rule->findOrFail(1);
+        $rounding = $rule->rounding;
+        $user_id = \Auth::user()->id;
+
+        $grand_total = \Cart::getTotal();
+        $number = rand(11111111, 99999999);
+        $date = \Carbon\Carbon::now();
+        if ($rounding) {
+            $grand_total = round($grand_total);
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:73]: make it ternary operator for $grand_total

File: app/Http/Controllers/FreeTrailController.php [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+                }
+            }
+        }
+
+        return 'success';
+    }
+
+    private function getIfFreetrailItemPresent($item, $invoiceid, $user_id, $order_status)
+    {
+        $product = Product::where('name', $item->product_name)->first()->id;
```
Comment (RafficMohammed) [inline app/Http/Controllers/FreeTrailController.php:128]: value('id')

File: app/Http/Controllers/FreeTrailController.php [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        ]);
+
+        return $invoiceItem;
+    }
+
+    private function executeFreetrailOrder()
+    {
+        $order_status = 'executed';
+        $userId = \Auth::user()->id;
+        $invoice = DB::table('invoices')->where('user_id', $userId)->first();
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:108]: use the model

File: app/Http/Controllers/FreeTrailController.php [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        $invoiceid = $invoice->id;
+
+        $invoice_items = DB::table('invoice_items')->where('invoice_id', $invoiceid)->get();
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:111]: same

File: app/Http/Controllers/FreeTrailController.php [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+                      case '3':
+                    $lastFour = '0'.$agents;
+                     break;
+                      case '4':
+                    $lastFour = $agents;
+
+                     break;
+                default:
+                    $lastFour = '0000';
+                    break;
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:187]: Don't have to use break for default case

File: app/Http/Controllers/Order/BaseOrderController.php [hunk 13]
```diff
@@ -154,7 +169,7 @@ public function addSubscription($orderid, $planid, $version, $product, $serial_k
             $licenseExpiry = $this->getLicenseExpiryDate($permissions['generateLicenseExpiryDate'], $days);
             $updatesExpiry = $this->getUpdatesExpiryDate($permissions['generateUpdatesxpiryDate'], $days);
             $supportExpiry = $this->getSupportExpiryDate($permissions['generateSupportExpiryDate'], $days);
-            $user_id = $this->order->find($orderid)->client;
+            $user_id = DB::table('orders')->find($orderid)->client;
```
Comment (sandesh556) [inline app/Http/Controllers/Order/BaseOrderController.php:172]: use the model

File: resources/views/themes/default1/layouts/front/master.blade.php [hunk 14]
```diff
@@ -121,6 +121,22 @@
                                        <div class="header-nav-main header-nav-main-effect-1 header-nav-main-sub-effect-1">
                                             <nav class="collapse">
                                                 <ul class="nav nav-pills" id="mainNav">
+
+                                                      @auth 
+                                                    <?php
+                                                    $id = \Auth::user()->id;
+                                                    $user = DB::table('users')->where('id', '=', $id)->value('first_time_login');?>
```
Comment (sandesh556) [inline resources/views/themes/default1/layouts/front/master.blade.php:128]: use model

File: routes/web.php [hunk 15]
```diff
@@ -97,6 +97,8 @@
              * Front Client Pages
              */
 
+            Route::post('firstlogin', 'FreeTrailController@firstloginatem');
```
Comment (sandesh556) [inline routes/web.php:100]: use - to maintain uniformity

File: app/Http/Controllers/FreeTrailController.php [hunk 16]
```diff
@@ -0,0 +1,199 @@
+<?php
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:1]: Also, use code comments I see no comments in most methods

File: app/Http/Controllers/FreeTrailController.php [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        $this->invoiceItem = $invoiceItem;
+
+        $order = new Order();
+        $this->order = $order;
+
+        $subscription = new Subscription();
+        $this->subscription = $subscription;
+    }
+
+    public function firstloginatem(Request $request)
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:38]: function names should be camelCase

File: app/Http/Controllers/FreeTrailController.php [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+            'currency' => $currency, ]);
+
+        return $invoice;
+    }
+
+    private function createFreetrailInvoiceItems()
+    {
+        $cart = \Cart::getContent();
+        $userId = \Auth::user()->id;
+        $invoice = DB::table('invoices')->where('user_id', $userId)->first();
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:86]: use model

File: app/Http/Controllers/FreeTrailController.php [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+        $product = Product::where('name', $item->product_name)->first()->id;
+        $version = Product::where('name', $item->product_name)->first()->version;
+        if ($version == null) {
+            //Get Version from Product Upload Table
+            $version = $this->product_upload->where('product_id', $product)->pluck('version')->first();
+        }
+        $serial_key = $this->generateFreetrailSerialKey($product, $item->agents); //Send Product Id and Agents to generate Serial Key
+        $domain = $item->domain;
+        //$plan_id = $this->plan($item->id);
+        $plan_id = Plan::where('product', '=', $product)
```
Comment (sandesh556) [inline app/Http/Controllers/FreeTrailController.php:137]: No need to use = in where clause

File: tests/Unit/FreeTrialControllerTest.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+use Illuminate\Support\Facades\Auth;
+use Tests\DBTestCase;
+
+class FreeTrailControllerTest extends DBTestCase
+{
+    public function test_firstloginatem_generateinvoiceorder_returnstatus200()
+    {
+        $this->assertFalse(auth()->check());
+        $user = factory(User::class)->create(['role' => 'user', 'country' => 'IN']);
+        $product = factory(Product::class)->create();
```
Comment (sandesh556) [inline tests/Unit/FreeTrialControllerTest.php:18]: no need of unnecessary variables

**Review by sandesh556 — CHANGES_REQUESTED** [review]

**Comment by sandesh556** [comment]
Also fix the codacy issues @sowmiyasubramaniyan

**Pull request closed without merging**